### PR TITLE
docs: add http guide sample and adjust text

### PR DESCRIPTION
**Pull request by wardbell**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

The current Http Guide does not have a working code sample. Worse, the code example for this guide is actually the code sample for the discontinued old HTTP module.

The guide does not reference that sample. Instead, it uses markdown ticks to present extracts from a hypothetical application. These extracts do not conform to style guidelines (e.g, they perform HTTP operations within the components instead of delegating to a service) and In several cases there are minor syntax errors.

## What is the new behavior?

Added a working `HttpClient` sample that is Hero themed and loosely based on ToH part 6 (the HTTP lesson).

The guide prose is adjusted to incorporate excerpts from the code sample. All substantive points in the original prose are retained.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by Foxandxss** [comment]
I added e2e but I can't make the upload example to work :/

**Comment by wardbell** [comment]
@Foxandxss Fixed the e2e by reworking the UploadInterceptor so it doesn't use `setInterval`.

File: aio/content/guide/http.md [hunk 1]
```diff
@@ -2,218 +2,397 @@
...
-Then, when you make the `HttpClient.get` call, pass a type parameter:
+The `HttpClient.get()` method parsed the JSON server response into the anonymous `Object` type. It doesn't know what the shape of that object is.
 
-```javascript
-http.get<ItemsResponse>('/api/items').subscribe(data => {
-  // data is now an instance of type ItemsResponse, so you can do this:
-  this.results = data.results;
-});
-```
+You can - and usually should - tell `HttpClient` the type of the response to make consuming the output easier and more obvious.
```
Comment (kapunahelewong) [inline aio/content/guide/http.md]: You can - and usually should - tell

^^ Don't forget `&mdash;` and to remove the spaces around them.

File: aio/content/guide/http.md [hunk 2]
```diff
@@ -2,218 +2,397 @@
...
+
+First define an interface with the correct shape:
```
Comment (kapunahelewong) [inline aio/content/guide/http.md]: Comma after "First".

**Review by kapunahelewong — CHANGES_REQUESTED** [review]
Looks great! Only a couple of small things.

File: aio/content/guide/http.md (hunk 2 above)
Comment (wardbell) [reply aio/content/guide/http.md]: Done

**Comment by wardbell** [comment]
The CI build failure [38478.2](https://travis-ci.org/angular/angular/jobs/327449737) - "gzip7 polyfills fell below expected size by >1%" is unrelated to this PR.

**Review by kapunahelewong — APPROVED** [review]

**Comment by petebacondarwin** [comment]
@wardbell - a rebase on master should solve your Travis failures.

**Review by alxhub — APPROVED** [review]
Looks good!

I do think we should keep the caching example - it demonstrates two critical things:

1) It's possible for an interceptor to respond to a request directly, without going to the server.
2) It's possible to send more than one response.

A simple `HttpCache` implementation could only look at URL (for GET requests of course).

If we can come up with a better example for the above two pieces, I'm all for it.

**Comment by wardbell** [comment]
@alxhub Interestingly, I demonstrate BOTH of those features in the [`HttpInterceptor#UploadInterceptor`](https://github.com/IdeaBlade/angular/blob/df3e0edad30e47958d3fbc097c165de708590536/aio/content/examples/http/src/app/http-interceptors.ts#L128) which the sample uses to simulate XHR upload behavior without having to create/configure a server to accept uploaded files.

Would that do?

**Comment by alxhub** [comment]
I say let's go with the npm search example. `UploaderInterceptor` does demonstrate the two points, but the mechanics involved are complicated and I think would distract from the concepts of interceptors themselves.

**Comment by wardbell** [comment]
@alxhub Will do!

**Comment by wardbell** [comment]
@alxhub @kapunahelewong The final `CachingInterceptor` (and corresponding prose) is done.
Added e2e test for it too. Looks ready to go.

File: aio/content/examples/http/src/app/http-interceptors/caching-interceptor.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+  intercept(req: HttpRequest<any>, next: HttpHandler) {
+    // continue if not cachable.
+    if (!isCachable(req)) { return next.handle(req); }
+
+    const cachedResponse = this.cache.get(req);
+    // #enddocregion v1
+    // #docregion intercept-refresh
+    // cache-then-refresh
+    if (req.headers.get('x-refresh')) {
```
Comment (alxhub) [inline aio/content/examples/http/src/app/http-interceptors/caching-interceptor.ts]: I think we should simplify the logic here and always make a request to refresh the cache, and return _both_ results to the user when there's also a previously cached value.

This also teaches an important property of interceptors - there is no need to limit yourself to one response.

**Review by alxhub — APPROVED** [review]

File: aio/content/examples/http/src/app/http-interceptors/caching-interceptor.ts (hunk 3 above)
Comment (wardbell) [reply aio/content/examples/http/src/app/http-interceptors/caching-interceptor.ts]: Note: Alex and I talked this through and he likes it as is.

**Review by IgorMinar — CHANGES_REQUESTED** [review]
Can we get the stackblitz PR #20165 merged first before this lands? otherwise the stackblitz PR will suffer another huge merge conflict.

**Comment by wardbell** [comment]
I'm cool with fitting this in after StackBlitz lands.  I anticipated that a bit by preparing the stackblitz.json files for this guide.

Sign me up for the conflict resolution when the time comes.

**Comment by wardbell** [comment]
The merge conflicts on Jan 24 are due to the Stackblitz merge. I will resolve and move to stackblitz at same time.

**Comment by wardbell** [comment]
Merge conflicts created by stackblitz PR resolved.
Stackblitz's for app and tests both working.
E2E tests and lint check pas.s
Docs are good and pass link check.

Ready to merge.

**Comment by wardbell** [comment]
The CI failure (as of this comment) appears to be an unrelated saucelabs flak.

**Review by petebacondarwin — APPROVED** [review]

**Pull request closed without merging**